zkevm-circuits/src/evm_circuit/execution/error_code_store.rs

```diff
@@ -0,0 +1,311 @@
+        let mut code = Bytecode::default();
+
+        // construct maxcodesize + 1 memory bytes
+        let code_creator: Vec<u8> = initialization_bytes
+            .to_vec()
+            .iter()
+            .cloned()
+            .chain(0u8..((32 - initialization_bytes.len() % 32) as u8))
+            .collect();
+        for (index, word) in code_creator.chunks(32).enumerate() {
+            code.push(32, Word::from_big_endian(word));
+            code.push(32, Word::from(index * 32));
+            code.write_op(OpcodeId::MSTORE);
+        }
+
+        if is_create2 {
+            code.append(&bytecode! {PUSH1(45)}); // salt;
+        }
+        code.append(&bytecode! {
+            PUSH32(initialization_bytes.len()) // size
+            PUSH2(0x00) // offset
+            PUSH2(23414) // value
+        });
+        code.write_op(if is_create2 {
+            OpcodeId::CREATE2
+        } else {
+            OpcodeId::CREATE
+        });
+        code.append(&bytecode! {
+            PUSH1(0)
+            PUSH1(0)
+            RETURN
+        });
+
+        code
+    }
+
+    fn test_context(caller: Account, is_oog: bool) -> TestContext<2, 1> {
+        TestContext::new(
+            None,
+            |accs| {
+                accs[0]
+                    .address(address!("0x000000000000000000000000000000000000cafe"))
+                    .balance(eth(10));
+                accs[1].account(&caller);
+            },
+            |mut txs, accs| {
+                txs[0]
+                    .from(accs[0].address)
+                    .to(accs[1].address)
+                    .gas(if is_oog {
+                        53800u64.into()
+                    } else {
+                        103800u64.into()
+                    });
+            },
+            |block, _| block,
+        )
+        .unwrap()
+    }
+
+    #[test]
+    fn test_create_codestore_oog() {
+        for is_create2 in [false, true] {
+            let initialization_code = initialization_bytecode(true);
+            let root_code = creator_bytecode(initialization_code, is_create2);
+            let caller = Account {
+                address: *CALLER_ADDRESS,
+                code: root_code.into(),
+                nonce: 1u64.into(),
+                balance: eth(10),
+                ..Default::default()
+            };
+            run_test_circuits(test_context(caller, true));
+        }
+    }
+
+    #[test]
+    fn test_create_max_code_size_exceed() {
+        for is_create2 in [false, true] {
+            let initialization_code = initialization_bytecode(false);
+            let root_code = creator_bytecode(initialization_code, is_create2);
+            let caller = Account {
+                address: *CALLER_ADDRESS,
+                code: root_code.into(),
+                nonce: 1u64.into(),
+                balance: eth(10),
+                ..Default::default()
+            };
+            run_test_circuits(test_context(caller, false));
+        }
+    }
+
```

Comment:
You missed a case that `max_code_size_exceed` and `codestore_oog` happen at the same time. Which maps to the constrain in L65.